### PR TITLE
App2App Authenticator addition to WSO2 OB accelerator

### PR DESCRIPTION
**Pull request by ImalshaD**
## App2App Authenticator addition to WSO2 OB accelerator

> Providing backend support for app2app flow in WSO2 OB3.

**Issue link:** *N/A*

**Doc Issue:** *N/A*

**Applicable Labels:** *Spec, product, version, type (specify requested labels)*

------

### Development Checklist

1. [x] Build complete solution with pull request in place.
2. [x] Ran checkstyle plugin with pull request in place.
3. [x] Ran Findbugs plugin with pull request in place.
4. [ ] Ran FindSecurityBugs plugin and verified report.
5. [x] Formatted code according to WSO2 code style.
6. [x] Have you verified the PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [x] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)?

### Testing Checklist

1. [x] Written unit tests.
2. [ ] Verified tests in multiple database environments (if applicable).
3. [x] Tested with BI enabled  (if applicable).

### CODE REVIEW

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 1]
```diff
@@ -46,6 +58,9 @@
  */
 public class JWTUtils {
 
+    private static final String DOT_SEPARATOR = ".";
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Shall we move this to OpenBankingConstants?

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 2]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+     * @param publicKey the public key that is used for validation
+     * @param algorithm the algorithm expected to have signed the jwt
+     * @return true if signature is valid else false
+     * @throws NoSuchAlgorithmException if the given algorithm doesn't exist
+     * @throws InvalidKeySpecException if the provided key is invalid
+     * @throws JOSEException if an error occurs during the signature validation process
+     */
+    @Generated(message = "Excluding from code coverage as KeyFactory does not initialize")
+    public static boolean validateJWTSignature(SignedJWT signedJWT, String publicKey, String algorithm) throws
+            NoSuchAlgorithmException, InvalidKeySpecException, JOSEException {
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]:
```suggestion
    public static boolean validateJWTSignature(SignedJWT signedJWT, String publicKey, String algorithm) 
            throws NoSuchAlgorithmException, InvalidKeySpecException, JOSEException {
```

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 3]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+
+        byte[] publicKeyData = Base64.getDecoder().decode(publicKey);
+        X509EncodedKeySpec spec = new X509EncodedKeySpec(publicKeyData);
+        KeyFactory kf = KeyFactory.getInstance(algorithm);
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:155]: Let's get the algorithm from JWT header

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 4]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+        RSAPublicKey rsapublicKey = (RSAPublicKey) kf.generatePublic(spec);
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:156]: Shall we check the possibility of using [this](https://www.javadoc.io/doc/com.nimbusds/nimbus-jose-jwt/5.1/com/nimbusds/jose/jwk/RSAKey.html#parse-java.lang.String-) method?

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 5]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+     *
+     * @param jwtString JWT string
+     */
+    public static boolean isJWT(String jwtString) {
+
+        if (jwtString == null) {
+            return false;
+        }
+
+        if (StringUtils.isBlank(jwtString)) {
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: This does the null check as well. Let's remove the previous if block

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 6]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+            return false;
+        }
+
+        return StringUtils.countMatches(jwtString, DOT_SEPARATOR) == 2;
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]:
```suggestion
        return StringUtils.isBlank(jwtString) ? false : StringUtils.countMatches(jwtString, DOT_SEPARATOR) == 2;
```

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 7]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+        JWSVerifier verifier = new RSASSAVerifier(rsapublicKey);
+        return signedJWT.verify(verifier);
+    }
+
+    /**
+     * Validate legitimacy of JWT.
+     *
+     * @param jwtString JWT string
+     */
+    public static boolean isJWT(String jwtString) {
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]:
```suggestion
    public static boolean isJWS(String jwsString) {
```

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 8]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+     * @throws IllegalArgumentException if the provided token identifier is not a parsable JWT
+     * Will not throw ParseException as it is already validated by isJWT
+     */
+    public static SignedJWT getSignedJWT(String jwtString) throws ParseException {
+
+        if (isJWT(jwtString)) {
+            return SignedJWT.parse(jwtString);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Provided token identifier is not a parsable JWT.");
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]:
```suggestion
                log.debug(String.format("Provided token identifier is not a parsable JWT: %s", jwtString));
```

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 9]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+            }
+            throw new IllegalArgumentException("Provided token identifier is not a parsable JWT.");
+        }
+    }
+
+    /**
+     * Validates whether a given JWT is not expired.
+     *
+     * @param expirationTime  the exp of the jwt that should be validated.
+     * @return true if the jwt is not expired
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:197]: missed defaultTimeSkew param. Applicable for other places as well

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java [hunk 10]
```diff
@@ -120,4 +135,101 @@ public static boolean validateJWTSignature(String jwtString, String jwksUri, Str
...
+     */
+    public static boolean validateExpiryTime(Date expirationTime, long defaultTimeSkew) {
```
Comment (VimukthiRajapaksha) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Applicable for other places as well

```suggestion
    public static boolean isValidExpiryTime(Date expirationTime, long defaultTimeSkew) {
```

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/test/java/com/wso2/openbanking/accelerator/common/test/util/JWTUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com).
```
Comment (amjadhifthikar) [inline open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/test/java/com/wso2/openbanking/accelerator/common/test/util/JWTUtilsTest.java]: Should be 2024. Check other places as well

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 10 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 9 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:197]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)
```suggestion
     * @return true if the jwt is not expired
```

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 4 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:156]: There is a issue when generating RSAKey using string the public key stored in db.

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 1 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 2 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 6 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/test/java/com/wso2/openbanking/accelerator/common/test/util/JWTUtilsTest.java (hunk 11 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/test/java/com/wso2/openbanking/accelerator/common/test/util/JWTUtilsTest.java]: Resolved in [7eaee49](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/7eaee4933908991557b77c90ffa5cb31dc2854c5)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 5 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 7 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

File: open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java (hunk 3 above)
Comment (ImalshaD) [reply open-banking-accelerator/components/com.wso2.openbanking.accelerator.common/src/main/java/com/wso2/openbanking/accelerator/common/util/JWTUtils.java:155]: Fixed in [e498dd5](https://github.com/ImalshaD/financial-open-banking/pull/6/commits/e498dd5122ae34f8806ded46629fdba1c0dd5fd7)

**Pull request merged**